snuba/clusters/cluster.py

```diff
@@ -8,6 +9,22 @@
 from snuba.reader import Reader, TQuery
 
 
+class ClickhouseClientSettings(Enum):
+    READONLY = {"readonly": True}
+    READWRITE: Mapping[str, Any] = {}
+    REPLACER = {
```

Comment:
I think this way of defining connection parameters seems quite ad-hoc for the specific use cases we have right now. I think we may be missing a decision to make:
given that the cluster is provided in configuration and the application should move towards a place where it does not care much about the connection. parameters, should connection settings be:
- hardcoded and classified in categories so that the code picks up the connection. it needs
- in settings and classified in categories so that the code keeps identifying the connection by a logical name, that then gets resolved to parameters
- entirely up to the application, such that the application code directly provides the connection settings it needs. 
(I don't like the third) but we should agree on the direction.

Also:
Do you have already an idea about whether we can transition the replacer to use reader and writer and not be Clickhouse specific ? I think we should have at least an idea on how to do that, because this ad-hoc connection seems inconsistent with the other two.
Looking at settings, having a generic readonly and readwrite makes sense, but mixing it with an application specific definition (replacer) makes it hard to understand which kinds of   connections are supposed to be added here.

On the other hand we could move totally the opposite way and have connection identifying the use case, like query, insert, replace, optimize, etc.
Which is the direction you plan to move towards ?

Reply:
I don't know of any reasons why most this these options will need to be configurable by users (apart from the replacer_max_block_size and replacer_max_memory_usage settings that already exists). Maybe in the future being able to adjust more of these options might be needed. Right now they seem largely application dependent to me. e.g. We'll always want readonly for querying and readwrite for DDL, etc.

Unfortunately, we can't do away with the replacer specific options and have generic read and readwrite only. So moving to the second option you mentioned (connection by use case) seems the most viable option to me.